src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,105 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int START_POSITION = 0;
+    private static final double CAPACITY_INCREASE = 1.5;
+    private int size;
+    private T[] elements;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public void ensureCapacity() {
+        if (size == elements.length) {
+            int newCapacity = (int) (size * CAPACITY_INCREASE);
+            T[] newElements = (T[]) new Object[newCapacity];
+            System.arraycopy(elements, START_POSITION, newElements, START_POSITION, size);
+            elements = newElements;
+        }
+    }
+
+    public void indexAddAvailable(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add value by " + index + " index");
+        }
+    }
+
+    public void indexAvailable(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("element with index " + index + " is not exist");
```

Comment:
Consider using a more informative exception message that includes the valid index range, for example: 'Index out of bounds: ' + index + ', valid range is [0, ' + (size - 1) + '].' This will make it clearer to the user what the valid range of indices is.